[expo-updates] add SelectionPolicyFilterAware

esamelson opened the pull request:
# Why

Support EAS Update's manifestFilters field, which is used for channels/rollouts. The client needs to be able to choose an update to launch, from a list of cached updates, without consulting the server--but using the stored value for `manifestFilters`. This PR adds a new `SelectionPolicy` on both platforms that takes any provided filters into account when selecting updates to load/launch/delete.

# How

Added `SelectionPolicyFilterAware`, which takes filters into account in the following ways:
- when selecting an update to launch, eliminates those which don't match the filters, then chooses the newest
- when selecting updates to delete, chooses to delete all updates that are older than the currently launched update, **except** the next newest update which matches the filters -- or, if no older updates match the filters, the next newest overall update
- when deciding whether to download a new update, uses the filters accordingly. There are some edge cases here which are better explained by reading the test cases. We will want to pass in the *new* filters here rather than any stored ones.

In this case, "match a filter" means that the manifest's `updateMetadata` object has a key that matches the filter **AND** the value doesn't match the expected one. In other words, if manifestFilters is the object `"channel": "rollout-1"`, an update with `updateMetadata: {channel: "default"}` would be filtered out but an update with `updateMetadata: {}` would **NOT** be filtered out.

# Test Plan

Added unit tests on both platforms for `SelectionPolicyFilterAware`.

# Todos

see https://www.notion.so/expo/EAS-Update-support-in-expo-updates-2178091dea9f4370aaf05f9a8ac328a6 . 
For this particular feature, next steps are to handle/parse the manifestFilters in the manifest response, persist them at the appropriate times, and then switch from `SelectionPolicyNewest` to `SelectionPolicyFilterAware`.

cc @jkhales

esamelson committed:
[android] add JSONObject filters parameter to all SelectionPolicy methods

esamelson committed:
[android] add SelectionPolicyFilterAware with tests

esamelson committed:
[ios] add filters parameter to all selection policy methods

esamelson committed:
[android] fix logic & tests

esamelson committed:
[ios] add EXUpdatesSelectionPolicyFilterAware and tests

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  @Test
+  public void testSelectUpdatesToDelete_OlderMatching() {
```
It may make sense to change this test to "SecondNewestMatching" and to add updateRollout3 to show that matching manifests still are deleted if they aren't the newest or second newest.

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingNewFilters() {
```
Do we want this behavior? What are its consequences? If the server sends down a manifest that doesn't match the filters, do we want the app to update to it?

esamelson replied:
That's a good point -- discussed with @jkhales and we should return false here, essentially sanity checking that a manifest passes its own filters before loading it (and persisting the filters).

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:
```diff
@@ -0,0 +1,147 @@
+
+  @Test
+  public void testIsUpdateManifestFiltered_Nested() throws JSONException {
+    JSONObject nestedManifestFilters = new JSONObject("{\"nested.object.key\": \"different-value\"}");
```
Using dots here precludes us from having dots inside of updateMetadata field names -- we need to update the spec if we are going to introduce this restriction.

esamelson replied:
Good point. Discussed with @jkhales and we decided that we're just going to stick with non-nested key names for now, as this seems to cover all reasonable cases and is fully supported by the existing structured fields spec.

ide commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:
```diff
@@ -0,0 +1,147 @@
+    // an update that doesn't have the key should not be filtered out
+    Assert.assertFalse(selectionPolicy.isUpdateManifestFiltered(updateDefault1, nestedManifestFilters));
+  }
+}
```
May also want to add a test that shows that if a manifest doesn't have a field the filter looks for, it's implicitly counted as a match. (That said, do we want that behavior? If a developer says to match a given field and then they send down a manifest that omits the field altogether, I don't think they'd expect it to run.)

esamelson replied:
Agree about adding a test. I'm pretty sure we do want this behavior; otherwise we'd need to start putting a `branchName` (or whatever other filters a user creates) into all embedded manifests in order for them not to be immediately filtered out.

(To be clear, I think we do want to allow users to specify `branchName` and other fields in embedded manifests, but we also don't want to force them to, which is what this "empty matches all" behavior allows.)

jkhales commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:
```diff
@@ -0,0 +1,147 @@
+
+  @Before
+  public void setup() throws JSONException {
+    manifestFilters = new JSONObject("{\"releaseName\": \"rollout\"}");
```
```suggestion
    manifestFilters = new JSONObject("{\"branchName\": \"rollout\"}");
```

We are renaming `release` to `branch`

esamelson replied:
thanks, changed (although the nice thing about this implementation is that it's totally agnostic of the actual keys inside of manifestFilters!)

esamelson committed:
remove support for nested keys; add unit tests for multiple filters and empty matches all

esamelson committed:
never load an update that doesn't match its own filters

esamelson committed:
add updateRollout0 to SecondNewestMatching test

esamelson committed:
releaseName -> branchName

esamelson commented:
Thanks for the feedback -- think I addressed all points.

Additionally, discussed with @ide and decided we want to **ignore** structured field value types that are not supported in JSON (e.g. binary) since there's no way we could have a manifest matching those values anyway; will add that functionality in a different PR since I think it makes most sense to do that when persisting the filters rather than when reading them.

esamelson committed:
update CHANGELOG
